schemachanger: support tbl zone config

annrpom opened the pull request:
### scbuild: add ResolvePhysicalTable
This patch adds `ResolvePhysicalTable` to
the name resolver interface. This will allow
us to set zone configurations on materialized
views and sequences -- which are both tables
that need to be stored in the kv layer.

Informs: https://github.com/cockroachdb/cockroach/issues/117574

Release note: None

---
### scop: remove unused SeqNum from AddDatabaseZoneConfig
The `SeqNum` is not needed during immediate mutation.

Release note: None

---
### scpb: modify TableZoneConfig element

This commit modifies the TableZoneConfig element
that will become relevant for supporting
`ALTER TABLE ... CONFIGURE ZONE` in the DSC.

Release note: None

---

### schemachanger: add table zone config opgen rule

schemachanger: add table zone config opgen rule
This patch adds an opgen rule for tbl zone configs
in the ADD direction.

Informs: https://github.com/cockroachdb/cockroach/issues/117574
Release note: None

---

### schemachanger: support tbl zone config

This patch adds the functionality to configure
a zone configuration on a table.

Fixes: https://github.com/cockroachdb/cockroach/issues/126495
Informs: https://github.com/cockroachdb/cockroach/issues/117574
Release note: None

Fixes: #129886

annrpom committed:
scpb: modify TableZoneConfig element

This commit modifies the TableZoneConfig element
that will become relevant for supporting
`ALTER TABLE ... CONFIGURE ZONE` in the DSC.

Informs: #117574
Release note: None

annrpom committed:
schemachanger: add table zone config opgen rule

This patch adds an opgen rule for tbl zone configs
in the ADD direction.

Informs: #117574
Release note: None

annrpom committed:
schemachanger: support tbl zone config

This patch adds the functionality to configure
a zone configuration on a table.

Fixes: #126495
Informs: #117574
Release note: None

separating for review

TODO squash into previous commit

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125786)
<!-- Reviewable:end -->

spilchen reviewed (commented):
Good timing with this review. The more of these DSC reviews I see, the more they are starting to make sense. I just had a few minor nits. Nothing major.

> Reviewed 8 of 8 files at r2, 5 of 5 files at r3, 15 of 15 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-:-O0FFxy56xBET7Lp4bbF:b3fyvci)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0F7NMHDphis3RO4qub:-O0F7NMHDphis3RO4quc:bvbzw1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L46)):*
> ```Go
> 
> const (
> 	databaseObj objType = iota
> ```

nit: suggest to start the enum with `unspecifiedObj`. That way this matches the default value.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 71 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0F85uHDrAly7SJX9om:-O0F85uHDrAly7SJX9on:b2ucywh) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L71)):*
> ```Go
> 	// - System Ranges
> 	var err error
> 	objectType, err = fallBackIfNotSupportedZoneConfig(n)
> ```

nit: any reason why `objectType` needs to be declared at the start of the function?



_Suggestion:_
```Go
	objectType, err := fallBackIfNotSupportedZoneConfig(n)
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FDMA603ictL0GjFcu:-O0FDMA603ictL0GjFcv:b6vscbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a3d8f7acc7c2f59b099f045c079895920511d7/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L91)):*
> ```Go
> 		panic(err)
> 	}
> 	elem.SeqNum = elem.SeqNum + 1
> ```

I couldn't see where the seqNum gets incremented now. Where is that?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 212 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FC7fY92uBQaYkkojf:-O0FC7fY92uBQaYkkojg:b-o8xvb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L212)):*
> ```Go
> 		// We're dealing with a table zone configuration change. Get the table descriptor so we can
> 		// determine if this is a multi-region table.
> 		tableID, err = getTargetIDFromZoneSpecifier(b, zs)
> ```

nit: any reason why we can't define `tableID` and `err` here?



_Suggestion:_
```Go
tableID, err := getTargetIDFromZoneSpecifier(b, zs)
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 939 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FESIaCxqq2S8OQpy1:-O0FESIaCxqq2S8OQpy2:bh5yglp) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L939)):*
> ```Go
> 
> // addZoneConfig adds the zone config to the build context.
> func addZoneConfig(
> ```

nit: at first glance, I didn't know the difference between this function and `applyZoneConfig`. Suggest another name. Perhaps `addZoneConfigToBuildCtx`?


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
> Reviewed 8 of 8 files at r2, 5 of 5 files at r3, 15 of 15 files at r5, 17 of 17 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 43 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fha9C7IL-4FkT9c2x:-O0Fha9DBUskt-NiBZep:b-pcupfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L43)):*
> ```Go
> // supported by the CONFIGURE ZONE statement. This is used to determine the
> // scpb that will be generated.
> type objType int
> ```

I think lets give it a zone config specific name, since it will be in the scbuildstmt package
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 121 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FmGD7EFNzamoov60G:-O0FmGD7EFNzamoov60H:b-g0ln4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L121)):*
> ```Go
> 	}
> 
> 	addZoneConfig(b, n, zc, seqNum, objectType)
> ```

Do we need to increment the sequence number?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FidhS6DliXHDr6Zhj:-O0FidhS6DliXHDr6Zhk:bumjbuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L271)):*
> ```Go
> // returned. Otherwise (for tables), the associated table ID is returned.
> func getTableIDFromZoneSpecifier(b BuildCtx, zs tree.ZoneSpecifier) (catid.DescID, error) {
> 	if zs.Database != "" {
> ```

Just have the fatal cases first, then you won't need a branch below:

```
if zs.Database != "" {
...
}

if !zs.TargetsTable() {
...
}
	tblName := zs.TableOrIndex.Table.ToUnresolvedObjectName()
		tableID := b.ResolveTable(tblName, ResolveParams{}).FilterTable().MustGetOneElement().TableID
		return tableID, nil


```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FiYTL8b9ElCp8w3cc:-O0FiYTL8b9ElCp8w3cd:b-n4pnq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L280)):*
> ```Go
> 		return tableID, nil
> 	}
> 	return 0, errors.AssertionFailedf("progrmaming error: zs does not specify a proper target")
> ```

Nit: programming
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 557 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fj6PC1aH2h-ir_FI5:-O0Fj6PC1aH2h-ir_FI6:bp7lsw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L557)):*
> ```Go
> 		}
> 	default:
> 		panic("programming error: unsupported object type for CONFIGURE ZONE")
> ```

errors.AssertionFailed
___
*[`pkg/sql/schemachanger/scop/immediate_mutation.go` line 902 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fg-CFE766huG8BokQ:-O0Fg-CFE766huG8BokR:b-qnom2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/184db29668e6ce689dc40cefe73cd7033d26a8e2/pkg/sql/schemachanger/scop/immediate_mutation.go#L902)):*
> ```Go
> 	TableID    descpb.ID
> 	ZoneConfig *zonepb.ZoneConfig
> 	SeqNum     uint32
> ```

We need this in the op? I don't think it gets used?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0F7NMHDphis3RO4qub:-O0P6Ojf0CX5f1aRTwqg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L46)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: suggest to start the enum with `unspecifiedObj`. That way this matches the default value.
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 71 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0F85uHDrAly7SJX9om:-O0P734k4Y_b1mJUa69a:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L71)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: any reason why `objectType` needs to be declared at the start of the function?
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FDMA603ictL0GjFcu:-O0Fjj-G0FzugbJg_JVG:b-mklyk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/45a3d8f7acc7c2f59b099f045c079895920511d7/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L91)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I couldn't see where the seqNum gets incremented now. Where is that?
</blockquote></details>

done -- facepalm. my hunch is that i accidentally deleted it during a rebase
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 212 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FC7fY92uBQaYkkojf:-O0P7W5D9WcWW2nSW-l6:b-cql6ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L212)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: any reason why we can't define `tableID` and `err` here?
</blockquote></details>

done -- remants of the past creeped up on me (read: refactoring)
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 939 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FESIaCxqq2S8OQpy1:-O0PCgwC1_w6z1RvkEvU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/947be5bffcc1cef16a7d1e8c8bec6dc27659d589/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L939)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: at first glance, I didn't know the difference between this function and `applyZoneConfig`. Suggest another name. Perhaps `addZoneConfigToBuildCtx`?
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 43 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fha9C7IL-4FkT9c2x:-O0P6Wcf-oEdQoGOo3-c:b-g7pros) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L43)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think lets give it a zone config specific name, since it will be in the scbuildstmt package
</blockquote></details>

done -- good point
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 121 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FmGD7EFNzamoov60G:-O0P78WsBVDaOMmf80lF:b-wzibzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L121)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we need to increment the sequence number?
</blockquote></details>

indeedly
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FidhS6DliXHDr6Zhj:-O0P81spA7CXlh7WWq8l:berz7po) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L271)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Just have the fatal cases first, then you won't need a branch below:

```
if zs.Database != "" {
...
}

if !zs.TargetsTable() {
...
}
	tblName := zs.TableOrIndex.Table.ToUnresolvedObjectName()
		tableID := b.ResolveTable(tblName, ResolveParams{}).FilterTable().MustGetOneElement().TableID
		return tableID, nil


```
</blockquote></details>

done -- i also realized that i am going to have to be more specific than targets table; so, i passed in my zoneConfigObjType type 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0FiYTL8b9ElCp8w3cc:-O0P9r-s8Nbnpe_wXfnb:b-7rl0jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L280)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: programming
</blockquote></details>

done 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 557 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fj6PC1aH2h-ir_FI5:-O0PA--rEUmV_v4yCWHC:bm2tba7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eff660aca3e070a3123ec9fc909edf744e54dee/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L557)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

errors.AssertionFailed
</blockquote></details>

done -- nice catch
___
*[`pkg/sql/schemachanger/scop/immediate_mutation.go` line 902 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Fg-CFE766huG8BokQ:-O0PCy3W8FTQBjrEg_Mm:b-5739v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/184db29668e6ce689dc40cefe73cd7033d26a8e2/pkg/sql/schemachanger/scop/immediate_mutation.go#L902)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We need this in the op? I don't think it gets used?
</blockquote></details>

oh fair point -- let me just add a commit that takes this out of db zone as well


<!-- Sent from Reviewable.io -->

annrpom committed:
scop: remove unused SeqNum from AddDatabaseZoneConfig

The `SeqNum` is not needed during immediate mutation.

Release note: None

spilchen reviewed (approved):
I had one minor suggestion and a couple of questions about the DSC. Nothing that needs to hold up the review. Looks good.

> Reviewed 27 of 27 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, 17 of 17 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-:-O0Q4DslE2fDbm8RLxEO:bd42ahe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 68 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q1RTx0J5Sf13OMn0X:-O0Q1RTy-pSuS0Effi9l:b-w45vkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L68)):*
> ```Go
> 	// - Partition/row
> 	// - System Ranges
> 	var err error
> ```

nit: I think we can remove this variable definition since we have a `:=` line for `err` below.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone/alter_table_configure_zone.explain` line 17 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q2mvl9S6StHjCcRuL:-O0Q2mvl9S6StHjCcRuM:b41sb2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone/alter_table_configure_zone.explain#L17)):*
> ```
>       ├── Stage 1 of 2 in PreCommitPhase
>       │    ├── 1 element transitioning toward PUBLIC
>       │    │    └── PUBLIC → ABSENT TableZoneConfig:{DescID: 104 (t), SeqNum: 0}
> ```

More of a general question, how are we able to apply this stage? When I look at the rules in `opgen_table_zone_config.go`, the PUBLIC->ABSENT transition is not implemented. It returns a `notImplementedForPublicObjects`.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain` line 10 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q3MhdBFiT-vvVPjoL:-O0Q3MhdBFiT-vvVPjoM:b-10eccq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain#L10)):*
> ```
> Schema change plan for ALTER TABLE ‹defaultdb›.‹public›.‹t› CONFIGURE ZONE USING ‹"gc.ttlseconds"› = ‹10000›; following ALTER TABLE ‹defaultdb›.‹public›.‹t› CONFIGURE ZONE USING ‹num_replicas› = ‹7›;
>  ├── StatementPhase
>  │    └── Stage 1 of 1 in StatementPhase
> ```

This is more a question about the DSC than your PR, but why is there nothing in this stage? The only difference between this test and an earlier one, which does perform an action here, is that this test updates an existing config. I assume this is related somehow.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> Reviewed 27 of 27 files at r7, 4 of 4 files at r8, 5 of 5 files at r9, 17 of 17 files at r10, 17 of 17 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @spilchen)

___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone/alter_table_configure_zone.explain` line 17 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q2mvl9S6StHjCcRuL:-O0UQQgH2_9MuYCGAF_y:b-uj5nex) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone/alter_table_configure_zone.explain#L17)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

More of a general question, how are we able to apply this stage? When I look at the rules in `opgen_table_zone_config.go`, the PUBLIC->ABSENT transition is not implemented. It returns a `notImplementedForPublicObjects`.
</blockquote></details>

So, the first stage of pre-commit isn't normal and its setup so that it can revert everything back with a single operation:  `UndoAllInTxnImmediateMutationOpSideEffects`. This doesn't use the opgen transitions at all and is a special case, if you look inside: buildStages, you'll see how its done.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain` line 10 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q3MhdBFiT-vvVPjoL:-O0USJyD4ttMdVsF54I-:b-fz0kna) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain#L10)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

This is more a question about the DSC than your PR, but why is there nothing in this stage? The only difference between this test and an earlier one, which does perform an action here, is that this test updates an existing config. I assume this is related somehow.
</blockquote></details>

This seems suspicious we should be able to observe side effects within the txn?  It doesn't matter today since we don't support txn's, but the current plan will mean that the changes are not "visible" inside the txn.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @spilchen)

___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain` line 10 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q3MhdBFiT-vvVPjoL:-O0UTdnQ6KSvGD3Ud8aE:bchydx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain#L10)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

This seems suspicious we should be able to observe side effects within the txn?  It doesn't matter today since we don't support txn's, but the current plan will mean that the changes are not "visible" inside the txn.
</blockquote></details>

Other plans seem fine, so I wonder if explains in this scenario have the wrong thing displayed?


<!-- Sent from Reviewable.io -->

annrpom committed:
scbuild: add ResolvePhysicalTable

This patch adds `ResolvePhysicalTable` to
the name resolver interface. This will allow
us to set zone configurations on materialized
views and sequences -- which are both tables
that need to be stored in the kv layer.

Informs: #117574

Release note: None

annrpom commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go:
```diff
@@ -504,17 +706,41 @@ func lookUpSystemZonesTable(
 				if e.DatabaseID == targetID {
 					zone = e.ZoneConfig
 					seqNum = e.SeqNum
+					// TODO(before merge) is this a small optimization we can do? DatabaseZoneConfig is on the
```
adding support for sequences and mat views rn (not too bad -- just have to do some more logic since it'll be broken up into 3 seperate elements)

also the answer to my question here: the answer is no since that's not how ForEach works -- i might be interested in making some sort of `func (c *ElementCollection[E]) Find(pred func(e E) bool) E` tho so we can short-circuit?

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 68 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q1RTx0J5Sf13OMn0X:-O0ihyWyBhpZz9pe5m44:b-7rl0jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L68)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: I think we can remove this variable definition since we have a `:=` line for `err` below.
</blockquote></details>

done 
___
*[`pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain` line 10 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-O0Q3MhdBFiT-vvVPjoL:-O0ik6RZ2MIn1ts69As3:be625e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c60954e57011a0ae1b97737bf1e2ce4ee475cf96/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple__statement_2_of_3.explain#L10)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Other plans seem fine, so I wonder if explains in this scenario have the wrong thing displayed?
</blockquote></details>

i had to rewrite these plans -- that would be my bad (i changed configuring range_min_bytes since validation prevents us from doing without setting range_max_bytes and i didnt want to generate another file)

i also had to rewrite alter_database_configure_zone_multiple for the same reason as i missed this validation in the db PR


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 2 of 30 files at r13, 33 of 34 files at r18, 2 of 2 files at r19, 4 of 4 files at r20, 5 of 5 files at r21, 32 of 32 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 30 files at r13, 34 of 34 files at r18, 2 of 2 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125786#-:-O1HW7fGESdvOMgd8jij:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

annrpom commented:
tftrs!

bors r+
